src/calculus.jl

```diff
@@ -0,0 +1,10 @@
+
+
+function diff(b1::Basic, b2::Basic)
```

Comment:
Sure, you can make Basic abstract and then subtype. But the issue you will
run into is you want symengine's computation to give the subclass
information, I doubt you want to do those computations in julia. I have
this outstanding issue (https://github.com/jverzani/SymPy.jl/issues/30)
that is related that I don't know how to solve. One help in SymPy is that
with PyCall, there is the `pytype_mapping` function that can be used to
specify a type that a python object should be coerced to on the julia side.
It isn't a precise as what I think you want, but better than what iI think
you have to work with here.

Rather, I would think you want to basically promote everything into a Basic
type and then have some means to convert back into a Julia value. I might
be wrong, but this is what i did with SymPy.jl.

On Fri, Mar 25, 2016 at 3:08 PM, Isuru Fernando <notifications@github.com>
wrote:

> In src/calculus.jl
> <https://github.com/symengine/SymEngine.jl/pull/17#discussion_r57478687>:
>
> > @@ -0,0 +1,10 @@
> > +
> > +
> > +function diff(b1::Basic, b2::Basic)
>
> Is it possible to have a type hierarchy in Basic, so that we can use diff(b1::Basic,
> b2::Symbol), where Symbol is a sub class of Basic?
>
> What we do in Python is that, we get the object and then depending on the
> runtime type, we create a Symbol or an Integer or an Add or a Mul or a Pow,
> etc which contains the object.
>
> How can we do this Julia?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/symengine/SymEngine.jl/pull/17/files/7343bd913def49c83b58a274d6363243ba1b9b57#r57478687>
>



-- 
John Verzani
Chair, Department of Mathematics
College of Staten Island, CUNY
verzani@math.csi.cuny.edu